tests: kernel: interrupt: Rework nested interrupt test

stephanosio opened the pull request:
The current nested interrupt test implementation is both buggy and
fundamentally flawed because it does not trigger a higher priority
interrupt from a lower priority interrupt context and relies on the
system timer interrupt, which is not fully governed by the test;
moreover, the current implementation does not properly validate the
test results and can report success if no interrupt is triggered and
serviced at all.

This commit reworks this test to have the following well-defined
and logical procedure:

1. [thread] Trigger IRQ 0 (lower priority)
2. [isr0] Set ISR 0 result token and trigger IRQ 1 (higher priority)
3. [isr1] Set ISR 1 result token and return
4. [isr0] Validate ISR 1 result token and return
5. [thread] Validate ISR 0 result token

The reworked test scenario ensures that the interrupt nesting works
properly and any abnormal conditions are detected (e.g. interrupts not
triggering at all, or ISR 1 not being nested under ISR 0).

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

Closes #23593

stephanosio committed:
tests: kernel: interrupt: Reorganise tests

This commit re-organises the kernel interrupt tests for consistency.

In addition, it removes any references to the `irq_offload` feature,
which is no longer used by this test.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: kernel: interrupt: Rework nested interrupt test

The current nested interrupt test implementation is both buggy and
fundamentally flawed because it does not trigger a higher priority
interrupt from a lower priority interrupt context and relies on the
system timer interrupt, which is not fully governed by the test;
moreover, the current implementation does not properly validate the
test results and can report success if no interrupt is triggered and
serviced at all.

This commit reworks this test to have the following well-defined
and logical procedure:

1. [thread] Trigger IRQ 0 (lower priority)
2. [isr0] Set ISR 0 result token and trigger IRQ 1 (higher priority)
3. [isr1] Set ISR 1 result token and return
4. [isr0] Validate ISR 1 result token and return
5. [thread] Validate ISR 0 result token

The reworked test scenario ensures that the interrupt nesting works
properly and any abnormal conditions are detected (e.g. interrupts not
triggering at all, or ISR 1 not being nested under ISR 0).

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
Output from the reworked test:
```
starting test - test_nested_isr
Triggering irq : 42
isr0: Enter
Triggering irq : 41
isr1: Enter
isr1: Leave
isr0: Leave
PASS - test_nested_isr
```

Output from the old (very broken) test:
```
starting test - test_nested_isr
isr0 running !!
Triggering irq : 41
isr1 ran !!
isr0 execution completed !!
Triggering irq : 42
isr0 running !!
isr0 execution completed !!
PASS - test_nested_isr
```

stephanosio commented:
NOTE: Shippable will fail for this test until https://github.com/zephyrproject-rtos/zephyr/issues/23593 is fixed, because the new test does not lie.

nashif commented:
> This commit reworks this test to have the following well-defined
> and logical procedure:
> 
>     1. [thread] Trigger IRQ 0 (lower priority)
> 
>     2. [isr0] Set ISR 0 result token and trigger IRQ 1 (higher priority)
> 
>     3. [isr1] Set ISR 1 result token and return
> 
>     4. [isr0] Validate ISR 1 result token and return
> 
>     5. [thread] Validate ISR 0 result token

Can you put this in the test documentation via  a doxygen header somewhere.

stephanosio committed:
tests: kernel: interrupt: Do not test interrupt nesting on RISC-V

This commit disables the nested interrupt test for the RISC-V platform,
as interrupt nesting is not supported on the current RISV-C
architecture port.

Furthermore, the current `trigger_irq` implementation for RISC-V is
mostly incorrect and cannot be used, so there is no point in leaving
that in the codebase (see #23593).

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
> Can you put this in the test documentation via a doxygen header somewhere.

@nashif I have added the documentation for `test_nested_isr`.

stephanosio commented:
Force pushing for further clean-up and adding documentation.

Also, I have added an additional commit that disables nested interrupt test for the RISC-V, as that seems to be not supported by the arch port and, even if it is supported, there is no practical way to test that in software because CLINT does not support priority-based interrupt nesting/preemption (for more details, see #23593).

stephanosio commented:
@npitre @kgugala @pgielda @nategraff-sifive It would be very helpful if one of the RISC-V maintainers could comment on the removal of the (previously broken and still is) nested interrupt testing for the RISC-V architectures.

ioannisg commented on tests/kernel/interrupt/prj.conf:
```diff
@@ -1,4 +1,3 @@
 CONFIG_ZTEST=y
-CONFIG_IRQ_OFFLOAD=y
```
This feature is not used in the test, correct?

stephanosio replied:
Correct. It used to be, but not anymore.

ioannisg commented on tests/kernel/interrupt/src/dynamic_isr.c:
```diff
@@ -0,0 +1,58 @@
+ */
+#if (defined(CONFIG_DYNAMIC_INTERRUPTS)		\
+	&& defined(CONFIG_GEN_SW_ISR_TABLE))
+#define DYNTEST 1
```
@stephanosio i wonder if you could clean up this local macro define - and use directly the corresponding CONFIG_ macro defines.

stephanosio replied:
Sure, I will address that in the PR for https://github.com/zephyrproject-rtos/zephyr/pull/23614#discussion_r399184649 as well.

ioannisg commented on tests/kernel/interrupt/src/nested_irq.c:
```diff
@@ -1,170 +1,133 @@
-#define ISR1_PRIO 1
+/*
+ * For Cortex-M NVIC, unused and available IRQs are automatically detected when
+ * when the test is run.
```
double "when"

stephanosio replied:
I will fix that in another PR since this already got merged.

stephanosio replied:
Addressed in #23881

ioannisg commented on tests/kernel/interrupt/src/nested_irq.c:
```diff
@@ -1,170 +1,133 @@
+ * and must be specified for every supported platform.
+ *
+ * In terms of priority, the IRQ1 is triggered from the ISR of the IRQ0;
+ * therefore, the priority of IRQ1 must be greater than that of the IRQ0.
```
"greater than" is this Cortex-specific requirement or applies to all? Just asking

stephanosio replied:
"greater" here means "greater in priority" and not "greater in numerical value," so it applies universally.

ioannisg commented on tests/kernel/interrupt/src/nested_irq.c:
```diff
@@ -1,170 +1,133 @@
+ * For Cortex-M NVIC, unused and available IRQs are automatically detected when
+ * when the test is run.
+ */
+#define IRQ0_PRIO	2
```
@stephanosio what is the ultimate reason that we need different priorities for Cortex-M?

stephanosio replied:
Mainly, I just copied what it was doing before, but the "ultimate reason" would be:
https://github.com/zephyrproject-rtos/zephyr/blob/685bf54829c9eb7a3a9b0667a4e0d1e9ddd27cfa/include/arch/arm/aarch32/exc.h#L44

But now that I am looking at it, it might have been better to do `_EXCEPTION_RESERVED_PRIO + N` here since the reserved priority offset is `1` instead of `0` if `CONFIG_CPU_CORTEX_M_HAS_PROGRAMMABLE_FAULT_PRIOS` is set.

stephanosio replied:
> it might have been better to do `_EXCEPTION_RESERVED_PRIO + N`

Actually, this is not necessary because `z_arm_irq_priority_set` already does that:
https://github.com/zephyrproject-rtos/zephyr/blob/685bf54829c9eb7a3a9b0667a4e0d1e9ddd27cfa/arch/arm/core/aarch32/irq_manage.c#L84

We should be okay as long as we do not use the priority value 0; I will add a comment about this in the upcoming PR.

stephanosio replied:
Addressed in #23881

ioannisg commented on tests/kernel/interrupt/src/interrupt_util.h:
```diff
@@ -55,19 +60,8 @@ static void trigger_irq(int irq)
 #endif
 }
 
-#elif defined(CONFIG_RISCV)
```
@stephanosio can you make a note that for RISCv the test does not run any more and link to the respective github issue?

stephanosio replied:
There is an issue about this:
https://github.com/zephyrproject-rtos/zephyr/issues/23593

As far as I can see, since this test should have never been enabled for RISC-V in the first place, I think this is really a non-issue.

https://github.com/zephyrproject-rtos/zephyr/issues/23593#issuecomment-605370756

ioannisg reviewed (commented):
Looks OK, I have some minor questions and comments for improvement
@stephanosio still the nested-test runs only for Cortex-M - not Cortex-R, correct?

carlocaione commented:
This was already tested on ARM64. ACK.

stephanosio commented:
> Looks OK, I have some minor questions and comments for improvement
> @stephanosio still the nested-test runs only for Cortex-M - not Cortex-R, correct?

@ioannisg This test runs on Cortex-R (and future AArch32 Cortex-A) as well, given that the system uses GIC (`qemu_cortex_r5` does).
